waves/src/ConfirmSwapDrawer.tsx

```diff
@@ -29,24 +29,78 @@ interface ConfirmSwapDrawerProps {
     onCancel: () => void;
     onSwapped: (txId: string) => void;
     transaction: string;
+    btcBalance: number;
+    usdtBalance: number;
 }
 
-export default function ConfirmSwapDrawer({ isOpen, onCancel, onSwapped, transaction }: ConfirmSwapDrawerProps) {
-    // eslint-disable-next-line @typescript-eslint/no-unused-vars
-    let { data, error } = useSWR("decompose-transaction", () => decomposeTransaction(transaction));
+export default function ConfirmSwapDrawer(
+    { isOpen, onCancel, onSwapped, transaction, btcBalance, usdtBalance }: ConfirmSwapDrawerProps,
+) {
+    const [alphaAsset, setAlphaAsset] = useState<AssetType>(AssetType.BTC);
+    const [alphaAmount, setAlphaAmount] = useState(0);
+    const [alphaBalanceBefore, setAlphaBalanceBefore] = useState(0);
+    const [alphaBalanceAfter, setAlphaBalanceAfter] = useState(0);
+    const [betaAsset, setBetaAsset] = useState<AssetType>(AssetType.USDT);
+    const [betaAmount, setBetaAmount] = useState(0);
+    const [betaBalanceBefore, setBetaBalanceBefore] = useState(0);
+    const [betaBalanceAfter, setBetaBalanceAfter] = useState(0);
+
+    function tickerToAssetType(ticker: string) {
+        switch (ticker) {
+            case LBTC_TICKER:
+                return AssetType.BTC;
+            case LUSDT_TICKER:
+                return AssetType.USDT;
+            default:
+                throw new Error("Unexpected ticker: " + ticker);
+        }
+    }
+
+    useAsyncEffect(async () => {
+        if (transaction !== "") {
+            let transactionBalance = await decomposeTransaction(transaction);
+
+            console.log(transactionBalance);
+
+            let n_assets = transactionBalance.length;
+
+            if (n_assets !== 2) {
+                console.log("Expected transaction involving 2 assets, got: " + n_assets);
+                throw new Error("Expected transaction involving 2 assets, got: " + n_assets);
+            }
+
+            // TODO: This assumes that one asset amount will be negative and one positive, which is probably a valid assumption forever, but we may want to explicitly check for this
+            transactionBalance.forEach(([ticker, amount]) => {
+                if (amount < 0) {
+                    let asset = tickerToAssetType(ticker);
+                    let balanceBefore = asset === AssetType.BTC ? btcBalance : usdtBalance;
+                    let balanceAfter = +balanceBefore + +amount;
+
+                    setAlphaAmount(-amount);
+                    setAlphaAsset(asset);
+                    setAlphaBalanceBefore(balanceBefore);
+                    setAlphaBalanceAfter(balanceAfter);
```

Comment:
I _think_ having several `set`'s is not really good for performance because it triggers multiple re-renders if React fails to batch them together.